src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -800,7 +801,14 @@ MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToDisk(
     {
         LOG_WARNING(log, "Directory {} already exists, probably result of a failed fetch. Will remove it before fetching part.",
             data_part_storage_builder->getFullPath());
-        data_part_storage_builder->removeRecursive();
+
+        /// Even if it's a temporary part it could be downloaded with zero copy replication and this function
+        /// is executed as a callback.
+        ///
+        /// We don't control the amount of refs for temporary parts so we cannot decide can we remove blobs
+        /// or not. So we are not doing it
+        bool keep_shared = disk->supportZeroCopyReplication() && data_settings->allow_remote_fs_zero_copy_replication;
+        data_part_storage_builder->removeSharedRecursive(keep_shared);
```

Comment:
Do we need similar fix in `clearOldTemporaryDirectories`?

Reply:
Yes